### ISSUE TITLE
Test Dependency Attribute

### ISSUE THREAD

**Issue by ashes999** [nunit/nunit#51]
Hi,

I have a web app with extensive automated testing. I have some installation tests (delete the DB tables and reinstall from scratch), upgrade tests (from older to newer schema), and then normal web tests (get this page, click this, etc.)

I switched from NUnit to MbUnit because it allowed me to specify test orders via dependency (depend on a test method or test fixture). I switched back to NUnit, and would still like this feature.

The current work-around (since I only use the NUnit GUI) is to order test names alphabetically, and run them fixture by fixture, with the installation/first ones in their own assembly.

**Comment by circa1741** [comment]
I have an idea that is more of a twist for dependency and ordering.

The discussion, so far, regarding dependency is "go ahead and run Test H (and possibly 8 other tests) only if Test 8 passes." In other words, there is no point of running Test H because if Test 8 fails then I know that Test H will also fail.

How about a dependency when a test fails?

Scenario:
I need a Smoke Test that covers a lot of ground. So, I am planning a Test Fixture that is an end-to-end test that has basic coverage of many of the SUT's features. The tests on said test fixture will use Test Ordering and are not "independent." The test order will be Test A then Test B then Test C, etc.

Now, because the tests are not "independent" I know that if Test C fails then all the following tests will also fail. Therefore, I need more tests to run in order to get a bigger picture of the Smoke Test.

I need to be able to configure to run Test 1 if Test A fails, Test 2 if Test B fails, Test 3 if Test C fails, etc.

My Test 3 is designed to be independent of Test B so if this fails then I have a better understanding of why Test C failed earlier. As it turns out my Tests 4 (for Test D), 5 (for E), 5 (for F), etc. all pass. Then I now understand that only the feature that was covered by Test C is the issue.

Why not run Tests 1, 2, 3, etc. instead? Well, because since those are isolated and independent tests then I am not doing Integration Tests. Again, I need a Smoke Test that covers a lot of ground.

Maybe something like:
- [DependsOnPassing("Test so and so")]
- [DependsOnFailing("Test blah blah blah")]

This will allow finer control in my automation test design.

**Comment by circa1741** [comment]
How about something like these attributes instead:
- [DependsOnPassing("Test so and so")]
- [RunOnFail("Test blah blah blah")]

Please note to which test these are attached to. These attributes should be useable in different levels (and in any): assembly, test fixture, test.

[DependsOnPassing("Test E")]
Test F
- Test E will automatically be executed if its result is unknown.
- Only then will Test F be determined if it should run or not.

[RunOnFail("Test N")]
Test I
- If Test I fails then Test N will automatically be executed.

Test N
- Test N may be executed on its own.
- But it will also automatically be executed if Test I fails.

Test E
- Test E may be executed on its own.
- But it will also automatically be executed if its result is unknown because Test F depends on this test passing.

**Comment by JeffCave** [comment]
Copied from #1031, which obviously duplicates this. Hopefully the example helps and the keywords direct others here... 

I have a couple of cases where I would like to mark a test as a prerequisite of another test. It would be nice to have an attribute that indicated tests that were prerequisites of the current test.

In the case where routines depend on one another, it is possible to know that a given test is going to fail because a sub-routine failed its test. If the test is a long running one, there really isn't any point in running the test if the sub-routine is broken anyway.

Contrived Example:

	public static class Statistics
	{
		public static double Average(IEnumerable<double> values)
		{
			double sum = 0;
			double count = 0;
			foreach (var v in values)
			{
				sum += v;
				count++;
			}
			return sum / count;
		}

		public static double MeanVariance(IEnumerable<double> values)
		{
			var avg = Average(values);
			var variance = new List<double>();
			foreach (var v in values)
			{
				variance.Add(Math.Abs(avg - v));
			}
			avg = Average(variance);
			return avg;
		}
	}

	[TestFixture]
	public class TestStatistics
	{
		[Test]
		public void Average()
		{
			var list = new List<double> { 1, 2, 3, 4, 5, 6, 7, 8, 9, 0 };
			var avg = Statistics.Average(list);
			Assert.AreEqual(4.5, avg);
		}

		[Test]
		//[Prerequisite("TestStatistics.Average")]
		public void MeanVariance()
		{
			//try { this.Average(); } catch { Assert.Ignore("Pre-requisite test 'Average' failed."); }
			var list = new List<double> { 1, 2, 3, 4, 5, 6, 7, 8, 9, 0 };
			var variance = Statistics.MeanVariance(list);
			Assert.AreEqual(0, variance);
		}
	}

Given the example, if the test `Average` fails, it makes sense to not bother testing `MeanVariance`.

I would conceive of this working by chaining the tests: 
* if `MeanVariance` is run, `Average` is forced to run first. 
* If `Average` has already been run, the results can be reused.
* If `Average` fails, `MeanVariance` is skipped.

**Comment by rladeira** [comment]
Are there any previsions when this feature will be available? I haven't found this information in others threads, maybe it is a duplicate question.

**Comment by CharliePoole** [comment]
It's in the 'Future' milestone, that means after 3.2, which is the latest actual milestone we have. However, we are about to reallocate issues to milestones, so watch for changes.

**Comment by fgsahoward** [comment]
In the spirit of extending NUnit to cover testing needs other than just unit tests, I'll weigh in on this conversation. The conversation so far has been a step in the right direction, but I feel there is a key aspect missing. 

Unit tests are inherently flat (i.e. setup, run one test, teardown), and should not have any dependency on any other test having been run. Integration testing adds the idea of ordered testing (or dependent tests for some definition of dependent) in which say the setup connects to the system, test A creates some object, test B modifies the object created by A, test C deletes the object created by A, and the tear-down disconnects from the system. This last example seems to be what most people are talking about in this thread. 

Consider UI testing of a website though, we have a page that can navigate to several other pages, and some actions can be performed on that page, which may affect the usage of other pages. In the unit testing structure, I need to navigate to page A, create object A.1, navigate to page B create object B.1. That's one test, which reimplements the test of creating object A.1 (from a different page). Now consider I need a test that modifies B.1 to point to A.1, I have to redo everything I just did to write this test. If we modify this example to fit the integration test analogy, I have to do slightly less work for the last step, in that I can rely on B.1 being in the same class and thus can do some sort of `[DependsOn("Create_B1_Object")]` attribute; however, if all the testing for page A is in a separate class, how could I link the two together?

The point here is that UI testing often involves a sort of tree structure for your test suite. All child tests under node B require that B completes successfully, and B depends on node A completing successfully, but neighbors to B do not depend on B being run at all. The tree structure allows you test all the functionality of page A independently of anything else, but since the layout and objects available on page B rely on some functionality of page A, it can depend on all those actions having been done already (by means of PageATests.CreateObject_A1 having been run as a parent).

In the tree structure, there is nothing that means you cannot have flat (unit) testing, or ordered (integration) testing, you can have the best of all of them, but you add on the ability to do UI testing.

Just my $.02 for what it's worth

**Comment by Sebazzz** [comment]
Exactly my thoughts. You'd want series of tests of which each test may have a dependency on the other. Series may or may not have a dependency on one another. More or less what is currently possible with MSTest Ordered Tests, other that MSTest also allows continuation of tests even if failure occurs.

A perfect example are UI tests which are often (always?) order and context sensitive. 

Other than MSTest there is no actively developed testing framework that allows developing such types of tests.

**Comment by CharliePoole** [comment]
@fgsahoward I'm in complete agreement with you at least at a high level - as a careful reading of past discussions will show. 

I say "careful" because I've disagreed with many specific details of proposals that have been made. As you can see, it is now listed as a feature on our backlog. In my view, we'll have to be careful when we implement it for two main reasons:

1. The feature is often requested by people writing unit tests.
2. Some folks (usually the same folks) mean something different by dependency than you do.

I'm generally of the opinion that true sequencing of tests, as a series of steps, is essential for many kinds of testing, incuding those you mention. I think we need to make that available in a specific way, that minimizes the chance folks will use them for other purposes. That probably means a different attribute being used to annotate the class or namespace that represents the overall test as well as another (different) attribute annotating the steps within that test. 

IOW, I don't want to stick this feature on top of TestFixture and Test.

The second issue will probably need to be dealt with by a dependency attribute of some kind, possibly including an indication of the relative "strength" of the dependency. IOW, if what you mean by dependency is "It's a waste of time to run B if A has failed, because it will certainly fail" then running test B or not can be at the option of the framework.

In fact, what you have described is what I call ordering. Ordering can be stricter than dependency. As a gui developer (my main focus for many years) I don't want to specify dependencies and have the framework figure out the order. I want to specify the ordering explicitly.

Another way I have talked about this in the past is to call what you are talking about "strict dependency" and the other kind "indirect dependency." That hasn't turned out to be very clear for most people. Perhaps when we produce some test implementations it will get clearer. :-)

**Comment by fgsahoward** [comment]
@Sebazzz Thanks for the tip about MsTest I may see if we can make that fit our needs in some fashion until this feature has been developed.

@CharliePoole I completely agree, there needs to be a very clear distinction between the types of tests you are about to start writing. I would not want all these different features rolled into one attribute with different configurations. I was more or less suggesting that the underlying implementation could use a tree structure for all types of tests, the attributes simply determine the ordering and dependencies.

I think it is very clear that no one would want to muddle the lines between the different styles of tests, so that you don't end up with a hack up of testing styles used for unit testing. Thank you for your notes, and I look forward to the solution you all come up with in the future. :)

**Comment by CharliePoole** [comment]
@oznetmaster Continuing the discussion from #170...

We have had extended discussions on the nunit lists in the past about what dependency means. Your comments would have fit in there quite well and describe what I have often called "strict dependency."

The thing is that we are driven by users and users who ask for dependency often mean something much less precise than what you describe. I did an analysis once and came up with three primary definitions of dependency. Like much of my past writing, it got lost in moving the project from one place to another, but if we are getting close to implementing something, I should try to recover it.

Basically - and possibly too simply - I found that people saying Test B depends on Test A meant one of three things:

1. Test A establishes pre-conditions that are needed for Test B to run or to succeed.
2. If test A fails, then test B will fail, so there is no need to run it. (different from 1 because the correlation is not causality, but due to both tests requiring the same feature in order to work.)
3. If test A fails, then that's all we need to know - test B will provide no useful information.

A key difference with types 2 and 3 is that the framework may choose to run them anyway if that's more convenient. With type 1, the framework should not run the Test B.

Dependencies in case of failure rather than success are a nice extension.

When we had this discussion, everyone with a view thought that there view was "the" definition of dependency and everyone else had to use a different word. :-) I'm less concerned with knowing the right definition than in giving each group what they are asking for.

**Comment by oznetmaster** [comment]
I fully agree with all of that.

However TestB runs after TestA only establishes an ordering, exactly as if TestA had an order of 1 and TestB had an order of 2.  There is no other implied dependency in the two tests.  TestB will run independent of TestA suceeding or failing, and there is no implicit assumption that TestA "sets up" TestB..  This is different then any of the three cases you listed.  This is why it is really an ordering semantic, not a dependency one.  It is exactly the same as using ordinals for ordering, but much more intuitive and easier to maintain and is self documenting.

**Comment by CharliePoole** [comment]
"TestB runs after TestA" is ambiguous. After TestA starts? finishes? succeeds?

However, I get what you are saying. It's possible to use a dependency syntax that simply translates into the sort of ordering we are doing. I suppose that could be a fourth meaning of "depends on" if we wanted to go there.

Are you suggesting we change what we are doing in this PR? Or do something else in future?

**Comment by CharliePoole** [comment]
@oznetmaster In the end we will likely have a max of one "OrderAttribute" (or similar) and one "DependsOnAttribute." We have to either pick a clear meaning for each one (if we have both) or provided some sort of property that allows the user to select from several meanings. Would you agree?

**Comment by oznetmaster** [comment]
Seems reasonable to me.  If we go with a dependency tree for all of them ,t he a single attribute would work, with different conditions for what its dependency means.

It seems that ordinal ordering is equally ambiguous.  What does ordering 1, 2, 3, actually mean.  Do they start in that order?  Do they wait for the earlier ones to finish before starting?

**Comment by oznetmaster** [comment]
I think that this PR could do it all.  Both ordering and dependency require building a logical dependency tree.  That is the first step.

What is done with that tree could be two PR's, or this single one.  I know you like smaller PRs, but I do not have a good grasp of how big doing it all in one PR would be.

Issues of cross TestFixture dependency (or TestFixture dependecy) also come into play, as do cross test assembly dependencies.  I am not suggesting we do them all, but we should at least consider them.  This applies to simple ordering as well.

**Comment by CharliePoole** [comment]
Except this isn't a PR. Not just a semantic quibble: no work has been done on it nore is it even scheduled or assigned. OTOH, the other one is something we have code for right now.

I think this is a pretty big deal unless some preliminary work is done separately first. We need to have an arbitrary way to specify dependencies between work items and schedule them accordingly. That's not how we do it now. We have no notion of items waiting for other items to finish. Issue #1096 could lead up to that. It's currently scheduled in milestone 3.4.

Of course, we could do some kind of adhoc dependency without that, but we would probably end up redoing it in the end. IMO, a generalized dependency graph between work items is what we really want to have. Then it becomes pretty easy to implement various user layers on top of it.

BTW, it feels like we often get into broad discussions like this at the point when we are trying to bring home a release target. I wish we could do it at the beginning of an iteration!

**Comment by oznetmaster** [comment]
I am very interested in working on this item.  I would start by proposing a "syntax" for specifying dependency.  Test ordering (non-dependent) would be part of it.  I would also like to make it work with the same syntax on TestFixture as well as Test.

As has been noted, this has been around for a long time.  Is it still too soon to actually start on this?

**Comment by CharliePoole** [comment]
No, not too soon to figure out the syntax anyway. I suggest you create a Specification on the dev wiki. I have some ideas that I would like to contribute... as very briefly outlined in one of the comments above. I think the key distinction is between "hard" dependencies, which NUnit must follow, and "softer" ones, which are basically just hints to the framework.

**Comment by oznetmaster** [comment]
I have a pretty good idea how to implement this as well, including using it to drive the SetUp and TearDown phases discussed in issue #1096.

It would mean changing the test "dispatching" from its current "push" mechanism to a "pull" one.

"Create a Specification on the dev wiki"?  No idea how to do that :(

**Comment by CharliePoole** [comment]
Not sure why you refer to test dispatching as a "push" mechanism. Workers pull items from the queues when they are ready to execute them. I'm planning to work on #1096 and I welcome any suggestions.

Can you edit a wiki? I'll create an empty page in the right place if you like. Otherwise, it can be text here, of course, but doing it on the wiki would give us a head start on documenting it later.

**Comment by oznetmaster** [comment]
CompositeWorkItem basically iterates across its children, and "pushes" each one to be executed.

I see an implementation of dependency which builds a linear "queue" of every work item in the test, sort of like a scheduler queue in an operating system, which specifies the conditions under which each item can be allowed to run.  The dispatcher removes the top "runable" work item from the queue to run (if parallel, then each work "thread" would remove the next runable work item that can be parallelized). When a work item completes, it then toggles the "runability" of other items in the queue, perhaps even discarding those that will now never be run due to their dependency.

It may even be possible to express parallelizability as a dependency condition.

I see #1096 as part and parcel of the same process.  Once multiple work items are created, they would each be assigned a dependency property which will control when they are executed.

**Comment by oznetmaster** [comment]
I have no idea if I cab edit a wiki.  Never tried, and have no idea how to.  Can you give me a starting "push"? :)

**Comment by rprouse** [comment]
@oznetmaster, editing the wiki is pretty easy, 

1. Pick a page you want to add a link to your new wiki page from, probably the [Specifications](https://github.com/nunit/dev/wiki/Specifications) page,
2. Edit the page by clicking the button
3. Add a link by surrounding text with double square brackets like `[[My NUnit Spec]]`
4. Save the page
5. Viewing the new link, it is red indicating the page does not exist
6. Click on the red link, it will take you to a create new page
7. Edit the page as you would an issue using GitHub markdown and save

**Comment by CharliePoole** [comment]
@oznetmaster What you describe is pretty much how the parallel dispatcher already works. Items are assigned to individual queues depending on their characteristics. Currently, it takes into account parallelizability and Apartment requirements. All items, once queued, are ready to be executed.

It was always planned that dependency queues would be added as a further step. I plan to use #1096 as an "excuse" to implement that infrastructure. Once it's implemented, it can then be further exposed to users as called for in  #51. I'll be preparing a spec for the underlying scheduling mechanism (dispatcher) as well and I'd like your comments on it.

**Comment by CharliePoole** [comment]
@oznetmaster I created an empty page for you: https://github.com/nunit/dev/wiki/Test-Dependency-Attribute

**Comment by oznetmaster** [comment]
Are we committed to call the attribute DependsOnAttribute?  How about something more general like "DependenciesAttribute"?

It is nigling, I know :(

**Comment by CharliePoole** [comment]
You should write it up as you think it should be. Then we'll all fight about it. :-)

**Comment by oznetmaster** [comment]
So I have :)

**Comment by CharliePoole** [comment]
Suggestion: add a section that explains motivation for each type of dependency. For example, when would a user typically want to use AfterAny, etc.

As a developer, it's always tempting to add things for "completeness." Trying to imagine a user needing each feature is a useful restraint on this tendency. Unfortunately, users generally only tell us when something is missing, not when something is not useful to them.

**Comment by Sebazzz** [comment]
For what its worth, my input:

I'd rather not define a dependency *per test method*, that is becoming rather tedious (and hard to maintain) if you have more than a few tests. Instead I want to establish an order between test fixtures. This comes from the following case we currently have: We are using MSTest for ordered tests currently. Except that it is MSTest, it works great, because with the test ordering I can express two thing about a test: Certain tests *may not execute before* another test. Other tests have a *dependency* on another test and may only *be executed after* the other test(s) have been executed.

Let's say that the integration test:
* Uses a test database with several user accounts in it
* The first few tests execute some tests using the test data, and also create test data themselves to be used in a later test. Note we have a *dependency* relationship here. Some tests may not execute if earlier tests fail.
* Then some browser-automation tests happen. They should be executed as late as possible, because they take a lot of time and we want to have feedback from earlier (faster) tests first.
* Finally, some logic is tested which deletes an entire user account. Note we have a *must not execute before* relationship here: If this test were to be done before the other tests, the other tests would fail.

With MSTest I can express this case fine: Each 'ordered test' in MSTest can contain ordered tests themselves. Also, ordered tests can have a flag set to abort if one of the tests fail.

```
             MyPrimaryOrderedTest
	          /      |         \
   DomainTests  BrowserTests  DestructiveTests
    /   |  \       /  |   \      |   \ 
   A    B   C     D   E    F     G    H 
```
For example, `MyPrimaryOrderedTest` has the fail 'abort on failure' set to false. There is nothing preventing `BrowserTests` to execute if `DomainTests` fail. However, `DomainTests` itself has the flag set to true so test C is not execute if A or B fail.

To be concrete, if was thinking of a interface like this to express the test fixture ordering:
```
interface ITestCollection {
    IEnumerable<Type> GetFixtureTypes();
    bool ContinueOnFailure { get; }
}
```
This is much more maintainable (and obvious) as having dependency attributes on every fixture and scales much better as the number of fixtures increase.

Note for test ordering *within* fixtures, I would simply use the existing `OrderAttribute` for that. I think test methods should not have inter-fixture test dependencies, because that makes the test structure too complex and unmaintainable.

**Comment by Sebazzz** [comment]
Update from my side: In the mean time I've managed to implement test ordering without the need to fork NUnit. It is "good enough" for me, so I use it now. It is already a lot better than the fragile state of many MSTest ordered tests.

**Comment by ravensorb** [comment]
Out of curiosity -- any chance the Dependency feature is planned for the next release?

**Comment by CharliePoole** [comment]
No plans at the moment. FYI, you can see that here on GitHub by virtue of the fact that it's not assigned to anyone and has no milestone specified.

For normal priority items, like this one, we don't usually pre-plan it for a particular release. We reserve that for high and critical items. This one will only get into a release plan when somebody decides they want to do it, self-assigns it and brings it to a point where it's ready for merging.

In fact, although not actually dependent on it, this issue does need a bunch of stuff from #164 to be effectively worked on. I'm working on that and expect to push it into the next release.

**Comment by Flynn1179** [comment]
Relevant: https://stackoverflow.com/questions/44112739/nunit-using-a-separate-test-as-a-setup-for-a-subsequent-test

Got referred here from there. I'm a firm believer that any fault should only ever cause one unit test to fail, having a whole bunch of others fail because they're dependent on that fault not being there is.. undesirable at best, and more than a little time consuming trying to track down which of the failing tests is the relevant one.

**Comment by espenalb** [comment]
@Flynn1179 - I agree with you when it comes to _unit_ tests. However, NUnit is also a great tool for other kinds of tests. For example, we use it for testing embedded firmware and are really missing this feature...

**Comment by CharliePoole** [comment]
@Flynn1179 Completely agree with you. There are techniques to prevent spurious failures such as you describe that don't "depend" on having a test dependency feature. In general, use an assumption to test those things that are actually tested in a different test and are required for your test to make sense.

It was a goal of NUnit 3 to extend NUnit for effective use with non-unit tests. We really have not done that yet - it may await another major release. OTOH, users are continuing to use it for those other purposes and trying to find clever ways to deal with the limitations. Here and there we have added small features and enhancements to help them, but it's really still primarily a unit-test framework.

Personally, I doubt I would want to use dependency as part of high-level testing myself. Rather, I'd prefer to have a different kind of fixture that executed a series of steps in a certain order, reporting success or failure of each step and either continuing or stopping based on some property of the step. That, however, is a whole other issue.

@espenalb I'd be interesting to know what you feel is needed particularly for testing embedded firmware.

**Comment by espenalb** [comment]
We are actually very happy about what NUnit offers. 

We use a combination of FixtureSetup/Setup/test attributes for configuring the device (Including flashing firmware)

Then we use different interfaces (serial port, jtag, ethernet) to interact with the devices, typically we send some commands and then observe results. Results can be command response, or in advanced tests we use dedicated hardware equipment for measuring device behavior.

The NUnit assertion macros, and FluentAssertions are then used to verify that everything is ok. 
By definition, these are all integration tests - and by nature a lot slower than a regular unit tests. The test dependency issue is therefore sorely missed - it is no point in verifying for example sensor performance if the command to enable sensor was rejected. The ability to pick up one test where another completed is therefore very valuable. 

With the test dependency attribute, we would have _one_ failing test, then _n_ ignored/skipped tests where the skipped tests could clearly state that they were not executed because the other tests failed...

Another difference from regular unit testing is heavy use of the log writer. There is one issue there regarding multithreading and log which I will create a separate issue for if it does not allready exist.

Bottom line from us - we are _very_ happy with NUnit as a test harness for integration testing. It gives us excellent support for a lot of advanced scenarios by using C# to interact with Device Under Test and other lab equipment.

With ReportUnit we then get nice html reports and we we also get Jenkins integration by using the rev2 nunit test output.

**Comment by ChrisMaddock** [comment]
> we also get Jenkins integration by using the rev2 nunit test output.

Complete aside, but the Jenkins plugin has recently been updated to read NUnit 3 output. 🙂

**Comment by DannyBraig** [comment]
Can someone give a small update about the status this feature? Is it planned?
In my department we are doing very long-running tests, which logically really depend on each other.
Some kind of "Test-Dependency" would be really interesting and helping for us...

I heard that you are in general planing to "open" NUnit for "non-unit test" tests as well (which is basically the case for us...). I think this attribute would be one step torwards it :-)

**Comment by Sebazzz** [comment]
This feature is still in design phase, so other than using external libraries there is no built-in support currently.

**Comment by aolszowka** [comment]
I am sorry to pull up an old thread but during the course of working though NUnit with a friend we stumbled into a case where if we had such a feature we could start to create integration tests (I realize NUnit is a Unit Testing Framework, but it seems like we could get what we want if we had Test Dependency).

First here's an updated link to the proposed spec (the link from CharliePoole here https://github.com/nunit/nunit/issues/51#issuecomment-188056417 was dead): https://github.com/nunit/docs/wiki/Test-Dependency-Attribute-Spec

Now for a use case; consider the following toy Program and Associated Tests

```csharp
namespace ExampleProgram
{
    using System.Collections;
    using NUnit.Framework;

    public static class ExampleClass
    {
        public static int Add(int a, int b)
        {
            return a - b;
        }

        public static int Increment(int a)
        {
            return Add(a, 1);
        }
    }

    public class ExampleClassTests
    {
        [TestCaseSource(typeof(AddTestCases))]
        public void Add_Tests(int a, int b, int expected)
        {
            int actual = ExampleClass.Add(a, b);
            Assert.That(actual, Is.EqualTo(expected));
        }

        [TestCaseSource(typeof(IncrementTestCases))]
        public void Increment_Tests(int a, int expected)
        {
            int actual = ExampleClass.Increment(a);
            Assert.That(actual, Is.EqualTo(expected));
        }
    }

    internal class IncrementTestCases : IEnumerable
    {
        public IEnumerator GetEnumerator()
        {
            yield return new TestCaseData(0, 1);
            yield return new TestCaseData(-1, 0);
            yield return new TestCaseData(1, 2);
        }
    }

    internal class AddTestCases : IEnumerable
    {
        public IEnumerator GetEnumerator()
        {
            yield return new TestCaseData(0, 0, 0);
            yield return new TestCaseData(0, 2, 2);
            yield return new TestCaseData(2, 0, 2);
            yield return new TestCaseData(1, 1, 2);
        }
    }
}
```

As an implementer I know that if any Unit Tests around `Add(int,int)` fail there is absolutely no point in running all the additional tests around `Increment(int)` other than noise. However there does not appear to be a way (short of Test Dependency) to specify this to NUnit (at least in my searches).

Doing a lot of research online it seems like others have worked around this by using a combination of factors (none of which are explicitly clear that `Increment(int)` depends on `Add(int,int)`) such ways include:

- Using Categories
- Using a Naming Convention To Control the Ordering of Tests

Neither of these seem to scale well, or even work for that matter when you use other features such as Parallel and all require some external "post processing" after the NUnit run has completed.

Is this the best path forward (if we were to use pure NUnit)? Is this feature still being worked on? (In other-words if a PR were submitted would it jam up anyone else working on something related?)

There is lots of good discussion in this thread about cyclic dependencies and other potential issues with this feature, it is obviously not easy to fix otherwise someone would have done it already. I am sure adding Parallel and TestCaseSource into the mix also increase complexity. I intend to dig more at some point, but before doing so wanted to make sure that this was not a solved problem or plans-in-the-works.

**Comment by CharliePoole** [comment]
@aolszowka 
This remains an accepted feature, at least as far as the issue labels go. @nunit/framework-team Am I correct there? 

Nobody has assigned it to themselves, which is supposed to mean that nobody is working on it. Smart of you to ask, none the less! If you want to work on it, some team member will probably assign it to themselves and "keep an eye" on you, since GitHub won't let us assign issues to non-members.

I made this a feature and gave it its "normal" priority back when I was project lead. I intended to work on it "some day" but never did and never will now that I'm not active in the project. I'm glad to correspond with you over any issues you find if you take it on.

My advice is to NOT do what I tried to do: write a complete spec and then work toward it. As you can read in the comments, we kept finding things to disagree about in the spec and nobody ever moved it to implementation. AFAIK (or remember) the prerequisite work in how tests are dispatched has been done already. I would pick one of the three types of dependency (see my two+ years ago comment) and just one use case and work on it. We won't want to release something until we are sure the API is correct, so you should probably count on a long-running feature branch that has to be periodically rebased or merged from master. Big job!

**Comment by ChrisMaddock** [comment]
> This remains an accepted feature, at least as far as the issue labels go. @nunit/framework-team Am I correct there?

Yes - as far as I'm concerned!

**Comment by Shiney** [comment]
Is anyone still interested in this being implemented? I might be interested in giving it a go?

**Comment by OsirisTerje** [comment]
It is still open, so feel free to give it a shot :-)

**Comment by OsirisTerje** [comment]
@Shiney Any possibility of doing this?

**Comment by Shiney** [comment]
Not in the short term, I'm planning on doing some parental leave at some point so if I end up having some free time to keep my C# up to date and this isn't yet done I will probably do this, but I have no free time at the moment.

**Comment by seba870701** [comment]
@Shiney  How did the parental leave go? Any possible updates on this one?

**Comment by Shiney** [comment]
Don't think I'll have time for this sorry

**Comment by stevenaw** [comment]
I've assigned this to myself. There have been many ideas proposed for this over time. I'll be pursuing implementing a minimal solution which can handle some of the more common use cases without constraining future enhancement. An MVP, in other words.

**Comment by jez9999** [comment]
In my particular case my use for this feature was performance tests - don't want to waste time running them if correctness fails.  In that case I guess it'd be handy to have them run only if every other test in the same fixture doesn't fail.

**Comment by CharliePoole** [comment]
@stevenaw It would be great to see something done on this.

I think one aspect of it is to decide which of the definitions of dependency you are intending to support. From past discussions and thinking, I recall that we considered the following...

1. A test may depend on another test must having already started. I think I called this pseudo-dependency some time in the past. It's what the `OrderAttribute` does when running tests in parallel.
2. A test may depend on another test having completed. It's what `OrderAttribute` already does when running tests without parallel execution.
3. A test must have completed successfully. That's what I think people are actually looking for. Fairly easy to do this for non-parallel execution, harder but needed for parallel.

The result criterion could vary. One variant that sounds useful to me is NOT failing. That would allow tests to run when they depended on a test that was inconclusive or dynamically ignored by use of Assert.Ignore.

Regarding @jez9999 ' comment, I'd lean towards running performance tests in a separate assembly, so there could be some argument for allowing assemblies to depend on other assemblies.

As an MVP, you could consider dependency among tests within a fixture, then work upward to Fixture dependency. I think mixed use cases -- like a method of one fixture depending on another fixture -- could easily be ignored for now.
